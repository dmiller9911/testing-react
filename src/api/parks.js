export function getParks() {
  return fetch('/parks.json').then(res => res.json());
}
