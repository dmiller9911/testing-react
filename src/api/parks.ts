export interface Park {
  name: string;
  location: string;
  maxElevation: number;
  minElevation: number;
  size: number;
  establishedDate: string;
  id: string;
}

export function getParks(): Promise<Park[]> {
  return fetch('/parks.json').then(res => res.json());
}
