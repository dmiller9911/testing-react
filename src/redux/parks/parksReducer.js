import {
  PARKS_FETCH_ALL_FAIL,
  PARKS_FETCH_ALL_REQUEST,
  PARKS_FETCH_ALL_SUCCESS,
  PARKS_TOGGLE_FAVORITE,
  PARKS_FILTER,
  parkFilters
} from './parksActions';

const defaultState = {
  byId: {},
  favorites: new Set(),
  error: null,
  isLoading: false,
  filter: parkFilters.all
};

export const stateKey = 'parks';

export function parksReducer(state = defaultState, action) {
  switch (action.type) {
    case PARKS_FETCH_ALL_REQUEST:
      return {
        ...state,
        isLoading: true
      };

    case PARKS_FETCH_ALL_SUCCESS:
      return {
        ...state,
        byId: action.payload.reduce(
          (newState, park) => ({
            ...newState,
            [park.id]: park
          }),
          {}
        ),
        error: null,
        isLoading: false
      };

    case PARKS_FETCH_ALL_FAIL:
      return {
        ...state,
        isLoading: false,
        error: action.payload
      };

    case PARKS_TOGGLE_FAVORITE:
      const parkId = action.payload.id;
      const favorites = new Set(state.favorites);
      if (favorites.has(parkId)) {
        favorites.delete(parkId);
      } else {
        favorites.add(parkId);
      }
      return {
        ...state,
        favorites
      };

    case PARKS_FILTER:
      return {
        ...state,
        filter: action.payload
      };

    default:
      return state;
  }
}
