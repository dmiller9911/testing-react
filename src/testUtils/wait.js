export function wait() {
  return new Promise(resolve => {
    setImmediate(resolve);
  });
}
