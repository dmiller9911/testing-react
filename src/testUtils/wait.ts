export function wait() {
  return new Promise(resolve => {
    setTimeout(resolve);
  });
}
