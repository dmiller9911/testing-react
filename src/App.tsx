import React, { useEffect } from 'react';
import { getParks, Park } from './api/parks';
import { ParkList } from './components/ParkList';
import { Container } from './components/Container';
import { Header } from './components/Header';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Switch from '@material-ui/core/Switch';
import { makeStyles } from './styles/makeStyles';
import { ParkFilter } from './components/ParkFilter';
import { createTestIds, getTestProps } from './utils/e2eUtils';

export const testIds = createTestIds('App', {
  showFavoritesOnlyToggle: 'showFavoritesOnlyToggle'
});

const useStyles = makeStyles(theme => ({
  root: {
    height: '100vh',
    display: 'grid',
    gridTemplateRows: 'max-content 1fr'
  },
  main: {
    overflowY: 'auto',
    backgroundColor: theme.palette.background.default,
    marginTop: -25,
    borderRadius: 25
  },
  filters: {
    display: 'grid',
    gridTemplateColumns: '1fr max-content',
    gridColumnGap: 50
  }
}));

export const App: React.FC = () => {
  const classes = useStyles();
  const [showFavoritesOnly, setShowFavoritesOnly] = React.useState(false);
  const [parkNameFilter, setParkNameFilter] = React.useState('');
  const [parks, setParks] = React.useState<Park[]>([]);
  const [favoriteParks, setFavoriteParks] = React.useState<Set<string>>(
    new Set()
  );

  const handleParkFavoriteToggle = React.useCallback((park: any) => {
    setFavoriteParks(currentValue => {
      const newValue = new Set(currentValue);
      if (newValue.has(park.id)) {
        newValue.delete(park.id);
      } else {
        newValue.add(park.id);
      }
      return newValue;
    });
  }, []);

  const handleParkNameFilterChange = React.useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      setParkNameFilter(e.target.value);
    },
    []
  );

  useEffect(() => {
    getParks().then(parks => {
      setParks(parks);
    });
  }, []);

  const filteredParks = React.useMemo(() => {
    if (showFavoritesOnly || parkNameFilter) {
      return parks.filter(p => {
        const isFavorite = !showFavoritesOnly || favoriteParks.has(p.id);
        return parkNameFilter
          ? new RegExp(parkNameFilter, 'gi').test(p.name) && isFavorite
          : isFavorite;
      });
    }
    return parks;
  }, [showFavoritesOnly, favoriteParks, parkNameFilter, parks]);

  return (
    <div className={classes.root}>
      <Header>
        <div className={classes.filters}>
          <ParkFilter
            label="Filter Parks by Name"
            value={parkNameFilter}
            onChange={handleParkNameFilterChange}
          />
          <FormControlLabel
            label="Favorites"
            style={{ color: '#fff' }}
            control={
              <Switch
                value="favoritesOnly"
                checked={showFavoritesOnly}
                inputProps={
                  getTestProps(testIds.showFavoritesOnlyToggle) as any
                }
                onChange={() => setShowFavoritesOnly(!showFavoritesOnly)}
              />
            }
          />
        </div>
      </Header>

      <main className={classes.main}>
        <Container>
          <ParkList
            parks={filteredParks}
            favoriteParks={favoriteParks}
            onToggleFavorite={handleParkFavoriteToggle}
          />
        </Container>
      </main>
    </div>
  );
};
