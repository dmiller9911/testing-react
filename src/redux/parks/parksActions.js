import { getParks } from '../../api/parks';

export const PARKS_FETCH_ALL_REQUEST = 'parks/fetch_all_request';
export const PARKS_FETCH_ALL_SUCCESS = 'parks/fetch_all_success';
export const PARKS_FETCH_ALL_FAIL = 'parks/fetch_all_fail';
export const PARKS_TOGGLE_FAVORITE = 'parks/toggle_favorite';
export const PARKS_FILTER = 'parks/filter';

export const parkFilters = {
  all: 'all',
  favorites: 'favorites'
};

export const getAllParksRequest = () => ({ type: PARKS_FETCH_ALL_REQUEST });
export const getAllParksSuccess = parks => ({
  type: PARKS_FETCH_ALL_SUCCESS,
  payload: parks
});
export const getAllParksFailure = err => ({
  type: PARKS_FETCH_ALL_FAIL,
  payload: err
});
export function getAllParks() {
  return dispatch => {
    dispatch(getAllParksRequest());
    getParks()
      .then(parks => dispatch(getAllParksSuccess(parks)))
      .catch(err => dispatch(getAllParksFailure(err)));
  };
}

export function toggleFavorite(park) {
  return {
    type: PARKS_TOGGLE_FAVORITE,
    payload: park
  };
}

export function setParkFilter(filter) {
  return {
    type: PARKS_FILTER,
    payload: filter
  };
}
