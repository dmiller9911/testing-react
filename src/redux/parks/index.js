import * as parksActions from './parksActions';
import * as parksSelectors from './parksSelectors';
import { parksReducer, stateKey as parksStateKey } from './parksReducer';

export { parksReducer, parksActions, parksSelectors, parksStateKey };
