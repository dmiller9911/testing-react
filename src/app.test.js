jest.mock('./api/parks');

import React from 'react';
import { shallow } from 'enzyme';
import App from './app';
import { getParks } from './api/parks';
import { wait } from './testUtils/wait';

const mockPark1 = createMockPark(1);
const mockPark2 = createMockPark(2);

getParks.mockReturnValue(Promise.resolve([mockPark1, mockPark2]));

it('renders', async () => {
  const view = shallow(<App />);
  await wait();
  view.update();
  expect(view).toMatchSnapshot();
});

function createMockPark(id) {
  return {
    id,
    name: `Park ${id}`
  };
}
