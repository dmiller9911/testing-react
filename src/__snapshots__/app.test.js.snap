// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="app_19tckwa"
>
  <Tabs
    onChange={[Function]}
    value={0}
  >
    <Tab
      label="All Parks"
      value={0}
    />
    <Tab
      label="Favorites"
      value={1}
    />
  </Tabs>
  <ParkList
    favoriteParks={Set {}}
    onToggleFavorite={[Function]}
    parks={
      Array [
        Object {
          "id": 1,
          "name": "Park 1",
        },
        Object {
          "id": 2,
          "name": "Park 2",
        },
      ]
    }
  />
</div>
`;
